src/humanloop/eval_utils/run.py

```diff
@@ -0,0 +1,739 @@
+from humanloop.types import ToolKernelRequest as Tool
+from humanloop.types.evaluation_run_response import EvaluationRunResponse
+from humanloop.types.run_stats_response import RunStatsResponse
+
+if typing.TYPE_CHECKING:
+    from humanloop.client import BaseHumanloop
+
+# Setup logging
+logger = logging.getLogger(__name__)
+logger.setLevel(level=INFO)
+console_handler = logging.StreamHandler()
+logger.setLevel(INFO)
+formatter = logging.Formatter("%(message)s")
+console_handler.setFormatter(formatter)
+if not logger.hasHandlers():
+    logger.addHandler(console_handler)
+
+EvaluatorDict = Union[CodeEvaluatorDict, LLMEvaluatorDict, HumanEvaluatorDict, ExternalEvaluator]
+Version = Union[FlowDict, PromptDict, ToolDict, EvaluatorDict]
+FileType = Literal["flow", "prompt", "tool", "evaluator"]
+
+
+# ANSI escape codes for logging colors
+YELLOW = "\033[93m"
+CYAN = "\033[96m"
+GREEN = "\033[92m"
+RED = "\033[91m"
+RESET = "\033[0m"
+
+
+CLIENT_TYPE = TypeVar("CLIENT_TYPE", PromptsClient, ToolsClient, FlowsClient, EvaluatorsClient)
+
+
+def log_with_evaluation_context(client: CLIENT_TYPE) -> CLIENT_TYPE:
+    """
+    Wrap the `log` method of the provided Humanloop client to use EVALUATION_CONTEXT.
+
+    This makes the overloaded log actions be aware of whether the created Log is
+    part of an Evaluation (e.g. one started by eval_utils.run_eval).
+    """
+
+    def _is_evaluated_file(
+        evaluation_context: EvaluationContext,
+        log_args: dict,
+        file_id_attribute: str,
+    ) -> bool:
+        """Check if the File that will Log against is part of the current Evaluation.
+
+        The user of the .log API can refer to the File that owns that Log either by
+        ID or Path. This function matches against any of them in EvaluationContext.
+        """
+        if evaluation_context == {}:
+            return False
+        return evaluation_context.get("file_id") == log_args.get(file_id_attribute) or evaluation_context.get(
+            "path"
+        ) == log_args.get("path")
+
+    # Copy the original log method in a hidden attribute
+    client._log = client.log  # type: ignore
+
+    def _overloaded_log(
+        self,
+        **kwargs,
+    ) -> Union[
+        CreatePromptLogResponse,
+        CreateToolLogResponse,
+        CreateFlowLogResponse,
+        CreateEvaluatorLogResponse,
+    ]:
+        evaluation_context: EvaluationContext
+        try:
+            evaluation_context = EVALUATION_CONTEXT.get() or {}  # type: ignore
+        except LookupError:
+            evaluation_context = {}  # type: ignore
+
+        if isinstance(client, PromptsClient):
+            file_id_attribute = "prompt_id"
+        elif isinstance(client, ToolsClient):
+            file_id_attribute = "tool_id"
+        elif isinstance(client, FlowsClient):
+            file_id_attribute = "flow_id"
+        elif isinstance(client, EvaluatorsClient):
+            file_id_attribute = "evaluator_id"
+
+        if _is_evaluated_file(
```

Comment:
this makes more sense to me now - this is what i was questioning yesterday when we were reviewing this method together 👍